integration/write_quorum_test.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/topology"
+	xlog "github.com/m3db/m3x/log"
+	xtime "github.com/m3db/m3x/time"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/shard"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNormalQuorum(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	// nodes = m3db nodes
+	nodes, closeFn, testWrite := makeNodes(t, []services.ServiceInstance{
+		node(t, 0, newClusterShardsRange(0, 1023, shard.Available)),
+		node(t, 1, newClusterShardsRange(0, 1023, shard.Available)),
+		node(t, 2, newClusterShardsRange(0, 1023, shard.Available)),
+	})
+
+	defer closeFn()
+
+	// Writes succeed to one node
+	require.NoError(t, nodes[0].startServer())
+	assert.NoError(t, testWrite(topology.ConsistencyLevelOne))
+	assert.Error(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.Error(t, testWrite(topology.ConsistencyLevelAll))
+
+	// Writes succeed to two nodes
+	require.NoError(t, nodes[1].startServer())
+	assert.NoError(t, testWrite(topology.ConsistencyLevelOne))
+	assert.NoError(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.Error(t, testWrite(topology.ConsistencyLevelAll))
+
+	// Writes succeed to all nodes
+	require.NoError(t, nodes[2].startServer())
+	assert.NoError(t, testWrite(topology.ConsistencyLevelOne))
+	assert.NoError(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.NoError(t, testWrite(topology.ConsistencyLevelAll))
+}
+
+func TestAddNodeQuorum(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	// nodes = m3db nodes
+	nodes, closeFn, testWrite := makeNodes(t, []services.ServiceInstance{
+		node(t, 0, newClusterShardsRange(0, 1023, shard.Leaving)),
+		node(t, 1, newClusterShardsRange(0, 1023, shard.Available)),
+		node(t, 2, newClusterShardsRange(0, 1023, shard.Available)),
+		node(t, 3, newClusterShardsRange(0, 1023, shard.Initializing)),
+	})
+
+	defer closeFn()
+
+	// No writes succeed to available nodes
+	require.NoError(t, nodes[0].startServer())
+	require.NoError(t, nodes[3].startServer())
+	assert.Error(t, testWrite(topology.ConsistencyLevelOne))
+	assert.Error(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.Error(t, testWrite(topology.ConsistencyLevelAll))
+
+	// Writes succeed to one available node
+	require.NoError(t, nodes[1].startServer())
+	assert.NoError(t, testWrite(topology.ConsistencyLevelOne))
+	assert.Error(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.Error(t, testWrite(topology.ConsistencyLevelAll))
+
+	// Writes succeed to two available nodes
+	require.NoError(t, nodes[2].startServer())
+	assert.NoError(t, testWrite(topology.ConsistencyLevelOne))
+	assert.NoError(t, testWrite(topology.ConsistencyLevelMajority))
+	assert.Error(t, testWrite(topology.ConsistencyLevelAll))
+}
+
+type testWriteFn func(topology.ConsistencyLevel) error
+
+func makeNodes(t *testing.T, instances []services.ServiceInstance) (testSetups,
```

Comment:
👍 👍 👍  parameterized cluster setup will definitely be useful down the line.

Want to refactor it so it doesn't need to return a `testWrite`, i.e. make a `func makeTestWrite(topoInit) testWriteFn { ... }` that you can call separately.

That way `makeNodes` can just focus on creating the nodes themselves?  Then we can reuse it in other tests down the line.  Once this refactor is done you can move `makeNodes(...)` to `integration.go` for reuse by others.

Also very minor nit: we could call it `newNodes(...)` to be slightly more consistent with our other constructor naming.